### PR TITLE
Add peerAddress to coyote request

### PR DESCRIPTION
**Pull request by rainerjung**
It contains the IP address of the direct connection peer.
If a reverse proxy sits in front of Tomcat and the protocol
used is AJP or HTTP in combination with the RemoteIp(Valve|Filter),
the peer address might differ from the remoteAddress.
The latter then contains the address of the client in front of the
reverse proxy, not the address of the proxy itself.

Support for the peer address has been added to the
RemoteAddrValve and RemoteCIDRValve with the new attribute
"usePeerAddress". This can be used to restrict access
to Tomcat bsed on the reverse proxy IP address, which is especially
useful to harden access to AJP connecrtors.

The peer address can also be logged in the access log
using the new %{peer}a syntax.

### CODE REVIEW

File: java/org/apache/catalina/valves/AbstractAccessLogValve.java [hunk 1]
```diff
@@ -861,19 +868,50 @@ public void addElement(CharArrayWriter buf, Date date, Request request,
      * write remote IP address - %a
      */
     protected class RemoteAddrElement implements AccessLogElement, CachedElement {
+        /**
+         * Type of address to log
+         */
+        private static final String remoteAddress = "remote";
```
Comment (martin-g) [inline java/org/apache/catalina/valves/AbstractAccessLogValve.java:874]: s/remoteAddress/REMOTE_ADDRESS/, to follow Java conventions ?! Same for `peerAddress` below.

File: java/org/apache/coyote/Constants.java [hunk 2]
```diff
@@ -96,4 +96,11 @@
      * the X-Forwarded-For HTTP header.
      */
     public static final String REMOTE_ADDR_ATTRIBUTE = "org.apache.tomcat.remoteAddr";
+
+    /**
+     * The request attribute set by the RemoteIpFilter, RemoteIpValve (and may
+     * be set by other similar components) that identifies for the connector the
+     * conection peer IP address.
```
Comment (martin-g) [inline java/org/apache/coyote/Constants.java:103]: con`n`ection

File: java/org/apache/coyote/Request.java [hunk 3]
```diff
@@ -95,6 +95,7 @@ public Request() {
 
     // remote address/host
     private final MessageBytes remoteAddrMB = MessageBytes.newInstance();
+    private final MessageBytes peerAddrMB = MessageBytes.newInstance();
```
Comment (martin-g) [inline java/org/apache/coyote/Request.java:98]: Do we care that it is either `remote` or `peer`? So one of them wastes memory. Or it is negligible ?

File: webapps/docs/changelog.xml [hunk 4]
```diff
@@ -117,6 +117,22 @@
...
+        the IP address of the direct connection peer. If a reverse proxy
+        sits in front of Tomcat and the protocol used is AJP or HTTP
+        in combination with the <code>RemoteIp(Valve|Filter)</code>,
+        the peer address might differ from the <code>remoteAddress</code>.
+        The latter then contains the address of the client in front of the
+        reverse proxy, not the address of the proxy itself.
+        Support for the peer address has been added to the
+        RemoteAddrValve and RemoteCIDRValve with the new attribute
+        <code>usePeerAddress</code>. This can be used to restrict access
+        to Tomcat bsed on the reverse proxy IP address, which is especially
```
Comment (martin-g) [inline webapps/docs/changelog.xml:131]: s/bsed/based/

File: webapps/docs/changelog.xml [hunk 5]
```diff
@@ -117,6 +117,22 @@
...
+        useful to harden access to AJP connecrtors. The peer address can also
```
Comment (martin-g) [inline webapps/docs/changelog.xml:132]: s/connecrtors/connectors/

File: java/org/apache/catalina/valves/AbstractAccessLogValve.java (hunk 1 above)
Comment (rainerjung) [reply java/org/apache/catalina/valves/AbstractAccessLogValve.java:874]: AbstractAccessLogValve consistently uses camel case for static Strings. I sticked to the local concentions in ther file. I would prefer to change the static string nameing conventions in an independent commit.

File: java/org/apache/coyote/Constants.java (hunk 2 above)
Comment (rainerjung) [reply java/org/apache/coyote/Constants.java:103]: Thanks.

File: java/org/apache/coyote/Request.java (hunk 3 above)
Comment (rainerjung) [reply java/org/apache/coyote/Request.java:98]: I think in many cases where one would need peer, remote can ba ueseful as well.

File: webapps/docs/changelog.xml (hunk 4 above)
Comment (rainerjung) [reply webapps/docs/changelog.xml:131]: Thanks.

File: webapps/docs/changelog.xml (hunk 5 above)
Comment (rainerjung) [reply webapps/docs/changelog.xml:132]: Thanks.

**Pull request merged**